protocols/mdns/src/behaviour/socket.rs

```diff
@@ -0,0 +1,149 @@
+        Self: Sized;
+
+    /// Attempts to receive a single packet on the socket from the remote address to which it is connected.
+    fn poll_receive_packet(
+        &mut self,
+        _cx: &mut Context,
+        _buf: &mut [u8],
+    ) -> Poll<Result<Option<(usize, SocketAddr)>, Error>>;
+
+    /// Attempts to send data on the socket to a given address.
+    fn poll_send_packet(
+        &mut self,
+        _cx: &mut Context,
+        _packet: &[u8],
+        _to: SocketAddr,
+    ) -> Poll<Result<(), Error>>;
+}
+
+#[cfg(feature = "async-io")]
+pub mod asio {
+    use super::*;
+    use async_io::Async;
+    use futures::FutureExt;
+
+    /// AsyncIo UdpSocket
+    pub type AsyncUdpSocket = Async<UdpSocket>;
+
+    impl AsyncSocket for AsyncUdpSocket {
+        /// Create the async socket from the ```std::net::UdpSocket```
+        fn from_socket(socket: UdpSocket) -> std::io::Result<Self> {
+            Async::new(socket)
+        }
+
+        /// Attempts to receive a single packet on the socket from the remote address to which it is connected.
+        fn poll_receive_packet(
+            &mut self,
+            cx: &mut Context,
+            buf: &mut [u8],
+        ) -> Poll<Result<Option<(usize, SocketAddr)>, Error>> {
+            // Poll receive socket.
+            let _ = futures::ready!(self.poll_readable(cx));
+            match self.recv_from(buf).now_or_never() {
+                Some(Ok((len, from))) => Poll::Ready(Ok(Some((len, from)))),
+                Some(Err(err)) => Poll::Ready(Err(err)),
+                None => Poll::Ready(Ok(None)),
+            }
+        }
+
+        /// Attempts to send data on the socket to a given address.
+        fn poll_send_packet(
+            &mut self,
+            cx: &mut Context,
+            packet: &[u8],
+            to: SocketAddr,
+        ) -> Poll<Result<(), Error>> {
+            let _ = futures::ready!(self.poll_writable(cx));
+            match self.send_to(packet, to).now_or_never() {
+                Some(Ok(_)) => Poll::Ready(Ok(())),
+                Some(Err(err)) => Poll::Ready(Err(err)),
+                None => Poll::Pending,
+            }
+        }
+    }
+}
+
+#[cfg(feature = "tokio")]
+pub mod tokio {
+    use super::*;
+    use ::tokio::net::UdpSocket as TkUdpSocket;
+
+    /// Tokio ASync Socket`
+    pub type TokioUdpSocket = TkUdpSocket;
+
+    impl AsyncSocket for TokioUdpSocket {
+        /// Create the async socket from the ```std::net::UdpSocket```
+        fn from_socket(socket: UdpSocket) -> std::io::Result<Self> {
+            socket.set_nonblocking(true)?;
+            TokioUdpSocket::from_std(socket)
+        }
+
+        /// Attempts to receive a single packet on the socket from the remote address to which it is connected.
+        fn poll_receive_packet(
+            &mut self,
+            cx: &mut Context,
+            buf: &mut [u8],
+        ) -> Poll<Result<Option<(usize, SocketAddr)>, Error>> {
+            match self.poll_recv_ready(cx) {
```

Comment:
You can use `?` here and also the `ready!` macro if you want.

```suggestion
            match futures::ready!(self.poll_recv_ready(cx)?) {
```